addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionHypersonicScanRule.java

```diff
@@ -267,146 +268,90 @@ public void init() {
      */
     @Override
     public void scan(HttpMessage originalMessage, String paramName, String paramValue) {
-        try {
-            // Timing Baseline check: we need to get the time that it took the original query, to
-            // know if the time based check is working correctly..
-            HttpMessage msgTimeBaseline = getNewMsg();
-            try {
-                sendAndReceive(msgTimeBaseline, false); // do not follow redirects
-            } catch (java.net.SocketTimeoutException e) {
-                // to be expected occasionally, if the base query was one that contains some
-                // parameters exploiting time based SQL injection?
-                LOGGER.debug(
-                        "The Base Time Check timed out on [{}] URL [{}]",
-                        msgTimeBaseline.getRequestHeader().getMethod(),
-                        msgTimeBaseline.getRequestHeader().getURI());
-            } catch (SocketException ex) {
-                LOGGER.debug(
-                        "Caught {} {} when accessing: {} for Base Time Check",
-                        ex.getClass().getName(),
-                        ex.getMessage(),
-                        msgTimeBaseline.getRequestHeader().getURI());
-                return; // No need to keep going
-            }
-            long originalTimeUsed = msgTimeBaseline.getTimeElapsedMillis();
-            // end of timing baseline check
 
-            int countUnionBasedRequests = 0;
-            int countTimeBasedRequests = 0;
+        LOGGER.debug(
+                "Scanning URL [{}] [{}], field [{}] with value [{}] for SQL Injection",
+                getBaseMsg().getRequestHeader().getMethod(),
+                getBaseMsg().getRequestHeader().getURI(),
+                paramName,
+                paramValue);
+
+        AtomicReference<HttpMessage> message = new AtomicReference<>();
+        AtomicReference<String> attack = new AtomicReference<>();
+        Iterator<String> it = SQL_HYPERSONIC_TIME_REPLACEMENTS.iterator();
+        for (int i = 0; !isStop() && it.hasNext() && i < blindTargetCount; i++) {
+            String sleepPayload = it.next();
+            TimingUtils.RequestSender requestSender =
+                    x -> {
+                        HttpMessage msg = getNewMsg();
+                        message.compareAndSet(null, msg);
+
+                        String finalPayload =
+                                sleepPayload
+                                        .replace(ORIG_VALUE_TOKEN, paramValue)
+                                        // Time in milliseconds for the SQL function.
+                                        .replace(SLEEP_TOKEN, Integer.toString(((int) x) * 1000));
+
+                        setParameter(msg, paramName, finalPayload);
+                        LOGGER.debug("Testing [{}] = [{}]", paramName, finalPayload);
+                        attack.compareAndSet(null, finalPayload);
```

Comment:
attack variable will always contains the first payload, even though attack payload is actually different other than 1st. I've confirmed this from my local testing.
@thc202

Reply:
Even though it is merged, looking forward to get this resolved.

Reply:
Right, the `AtomicReference`s should be declared inside the loop.

Reply:
#5163